src/analyses/cfg_dominators.h

```diff
@@ -142,96 +293,201 @@ void cfg_dominators_templatet<P, T, post_dom>::initialise(P &program)
+  std::size_t entry_node_index) const
+{
+  auto reverse_postordering = [this](T lhs, T rhs) {
+    return cfg.get_node(lhs).postorder_index >
+           cfg.get_node(rhs).postorder_index;
+  };
+
+  std::vector<T> order;
+
+  // Note that this will only select and order nodes that have been processed.
+  //
+  for(std::size_t i = 0; i < cfg.size(); ++i)
+  {
+    if(
+      i != entry_node_index &&
+      cfg[i].postorder_index != nodet::NODE_NOT_VISITED)
+    {
+      order.push_back(cfg[i].PC);
+    }
+  }
+
+  std::sort(order.begin(), order.end(), reverse_postordering);
+
+  return order;
+}
+
 /// Computes the MOP for the dominator analysis
 template <class P, class T, bool post_dom>
 void cfg_dominators_templatet<P, T, post_dom>::fixedpoint(P &program)
 {
-  std::list<T> worklist;
-
   if(cfgt::nodes_empty(program))
     return;
 
+  // Initialise entry node to be its own dominator:
   if(post_dom)
     entry_node = cfgt::get_last_node(program);
   else
     entry_node = cfgt::get_first_node(program);
   typename cfgt::nodet &n = cfg.get_node(entry_node);
-  n.dominators.insert(entry_node);
+  const auto entry_node_index = cfg.get_node_index(entry_node);
+  n.dominator = entry_node_index;
 
-  for(typename cfgt::edgest::const_iterator
-      s_it=(post_dom?n.in:n.out).begin();
-      s_it!=(post_dom?n.in:n.out).end();
-      ++s_it)
-    worklist.push_back(cfg[s_it->first].PC);
+  // Calculate a post-ordering on the CFG:
+  assign_postordering(entry_node_index);
 
-  while(!worklist.empty())
-  {
-    // get node from worklist
-    T current=worklist.front();
-    worklist.pop_front();
+  const auto visit_order =
+    get_reverse_postordered_instructions(entry_node_index);
 
-    bool changed=false;
-    typename cfgt::nodet &node = cfg.get_node(current);
-    if(node.dominators.empty())
-    {
-      for(const auto &edge : (post_dom ? node.out : node.in))
-        if(!cfg[edge.first].dominators.empty())
-        {
-          node.dominators=cfg[edge.first].dominators;
-          node.dominators.insert(current);
-          changed=true;
-        }
-    }
+  bool any_change = true;
 
-    // compute intersection of predecessors
-    for(const auto &edge : (post_dom ? node.out : node.in))
+  while(any_change)
+  {
+    any_change = false;
+    for(const auto current : visit_order)
     {
-      const target_sett &other=cfg[edge.first].dominators;
-      if(other.empty())
-        continue;
-
-      typename target_sett::const_iterator n_it=node.dominators.begin();
-      typename target_sett::const_iterator o_it=other.begin();
+      auto &node = cfg.get_node(current);
+      const cfg_nodet *dominator_node = nullptr;
 
-      // in-place intersection. not safe to use set_intersect
-      while(n_it!=node.dominators.end() && o_it!=other.end())
+      // compute intersection of predecessors
+      auto &edges = (post_dom ? node.out : node.in);
+      if(edges.size() != 0)
```

Comment:
This check is still pointless (the for-loop will terminate immediately if there are no edges)

Reply:
Caused exceptions at one point when attempting to get the in/out iterators, hopefully not the case anymore.